compiler/rustc_lexer/src/lib.rs

```diff
@@ -277,6 +277,62 @@ pub fn strip_shebang(input: &str) -> Option<usize> {
     None
 }
 
+/// Frontmatter is a special attribute type reserved for use by external tools
+///
+/// This must be called after [`strip_shebang`]
+pub fn strip_frontmatter(input: &str) -> Option<usize> {
+    // Whitespace may precede a frontmatter but must end with a newline
+    let mut rest = input;
+    while !rest.is_empty() {
+        let without_spaces = rest.trim_start_matches(|c| is_whitespace(c) && c != '\n');
+        let without_nl = without_spaces.trim_start_matches('\n');
+        if without_nl.len() == rest.len() {
+            // nothing trimmed
+            break;
+        } else if without_nl.len() == without_spaces.len() {
+            // either not a frontmatter or invalid opening
+            return None;
+        }
+        rest = without_nl;
+    }
+
+    // Opens with a line that starts with 3+ `-` followed by an optional identifier
+    const FENCE_CHAR: char = '-';
+    let fence_end =
+        rest.char_indices().find_map(|(i, c)| (c != FENCE_CHAR).then_some(i)).unwrap_or(rest.len());
+    if fence_end < 3 {
+        // either not a frontmatter or invalid frontmatter opening
+        return None;
+    }
+    let (fence_pattern, rest) = rest.split_at(fence_end);
+    let (info, rest) = rest.split_once("\n").unwrap_or((rest, ""));
+    let info = info.trim_matches(is_whitespace);
```

Comment:
```suggestion
    let info = info.trim();
```

Reply:
Is that a hard invariant that `rustc_lexer::is_whitespace` and `char::is_whitespace` match?  Otherwise, I'd feel more comfortable using the same definition as the rest of the Rust lexer.